modules/gdscript/gdscript_analyzer.cpp

```diff
@@ -1358,27 +1366,38 @@ void GDScriptAnalyzer::resolve_class_body(GDScriptParser::ClassNode *p_class, co
 
 					if (setter_function == nullptr) {
 						push_error(vformat(R"(Setter "%s" not found.)", member.variable->setter_pointer->name), member.variable);
-
-					} else if (setter_function->parameters.size() != 1) {
-						push_error(vformat(R"(Function "%s" cannot be used as setter because of its signature.)", setter_function->identifier->name), member.variable);
-
-					} else if (!is_type_compatible(member.variable->datatype, setter_function->parameters[0]->datatype, true)) {
-						push_error(vformat(R"(Function with argument type "%s" cannot be used as setter for a property of type "%s".)", setter_function->parameters[0]->datatype.to_string(), member.variable->datatype.to_string()), member.variable);
-
 					} else {
-						has_valid_setter = true;
-
+						if (setter_function->mandatory_parameter_count == 0 || setter_function->mandatory_parameter_count > 2) {
+							push_error(vformat(R"(Setters must have 1 or 2 mandatory arguments.)"), member.variable);
+						} else {
+							GDScriptParser::ParameterNode *value_arg = setter_function->parameters[setter_function->mandatory_parameter_count - 1];
+							if (!is_type_compatible(member.variable->datatype, value_arg->datatype, true)) {
+								push_error(vformat(R"(Value argument of type "%s" is not compatible with value type "%s".)", value_arg->datatype.to_string(), member.variable->datatype.to_string()), member.variable);
+							} else {
+								has_valid_setter = true;
+								if (setter_function->mandatory_parameter_count == 2) {
+									GDScriptParser::DataType name_arg_type = setter_function->parameters[0]->datatype;
+									if (!name_arg_type.is_variant() && name_arg_type.is_hard_type()) {
+										if (name_arg_type.kind != GDScriptParser::DataType::BUILTIN || (name_arg_type.builtin_type != Variant::STRING && name_arg_type.builtin_type != Variant::STRING_NAME)) {
+											push_error(vformat(R"(First mandatory argument of setters with 2 mandatory arguments must be a property name of type String or StringName.)"), member.variable);
```

Comment:
```suggestion
											push_error(R"(First mandatory argument of a setter with 2 mandatory arguments must be String or StringName.)", member.variable);
```

Reply:
Agree mostly here, but I prefer "must be of type String or StringName" to "must be String or StringName" because the latter could be interpreted as meaning the argument literally has to be one of two values: `String` or `StringName`.